File: src/summary.js
// Aggregation shared by GET /expenses (live summary) and the monthly report.

import { displayFromDate } from "./dates.js";

const round2 = (n) => Math.round((n + Number.EPSILON) * 100) / 100;

// "DD-MM-YYYY" -> sortable number YYYYMMDD
const sortable = (dmy) => {
  const [d, m, y] = dmy.split("-");
  return Number(`${y}${m}${d}`);
};

/** Calendar days in a "YYYY-MM". */
export function daysInMonth(ym) {
  const [y, m] = ym.split("-").map(Number);
  return new Date(Date.UTC(y, m, 0)).getUTCDate();
}

/**
 * Compute the full summary for a set of expense docs over `dayCount` days.
 * Returns numeric fields only; callers add any labels (month/range).
 */
export function computeSummary(expenses, dayCount = 1) {
  const grandTotal = round2(expenses.reduce((s, e) => s + (e.amount || 0), 0));
  const transactionCount = expenses.length;
  const avgPerDay = round2(grandTotal / Math.max(1, dayCount));

  // by category
  const catMap = new Map();
  for (const e of expenses) {
    const c = e.category || "Other";
    const cur = catMap.get(c) || { total: 0, count: 0 };
    cur.total += e.amount || 0;
    cur.count += 1;
    catMap.set(c, cur);
  }
  const breakdown = [...catMap.entries()]
    .map(([category, v]) => ({
      category,
      total: round2(v.total),
      count: v.count,
      percentage: grandTotal > 0 ? round2((v.total / grandTotal) * 100) : 0,
    }))
    .sort((a, b) => b.total - a.total);

  // by source (only need the leader)
  const srcMap = new Map();
  for (const e of expenses) {
    const s = e.source || "Other";
    srcMap.set(s, (srcMap.get(s) || 0) + (e.amount || 0));
  }
  const topSource = [...srcMap.entries()].sort((a, b) => b[1] - a[1])[0]?.[0] || "—";
  const topCategory = breakdown[0]?.category || "—";

  // daily trend
  const dayMap = new Map();
  for (const e of expenses) {
    const d = e.displayDate || displayFromDate(e.date);
    dayMap.set(d, (dayMap.get(d) || 0) + (e.amount || 0));
  }
  const dailyTrend = [...dayMap.entries()]
    .map(([date, total]) => ({ date, total: round2(total) }))
    .sort((a, b) => sortable(a.date) - sortable(b.date));

  return {
    grandTotal,
    transactionCount,
    avgPerDay,
    topCategory,
    topSource,
    breakdown,
    dailyTrend,
  };
}

File: android/app/src/main/java/com/anuj/expensetracker/data/model/Summary.kt
package com.anuj.expensetracker.data.model

import kotlinx.serialization.Serializable

@Serializable
data class CategoryBreakdown(
    val category: String,
    val total: Double,
    val count: Int,
    val percentage: Double,
)

@Serializable
data class DailyTrendPoint(
    // "DD-MM-YYYY"
    val date: String,
    val total: Double,
)

@Serializable
data class Summary(
    val grandTotal: Double,
    val transactionCount: Int,
    val avgPerDay: Double,
    val topCategory: String,
    val topSource: String,
    val breakdown: List<CategoryBreakdown> = emptyList(),
    val dailyTrend: List<DailyTrendPoint> = emptyList(),
)

@Serializable
data class DateRange(
    val start: String,
    val end: String,
)

@Serializable
data class ExpensesResponse(
    val month: String? = null,
    val displayMonth: String? = null,
    val from: String,
    val to: String,
    val days: Int,
    val range: DateRange,
    val summary: Summary,
    val expenses: List<Expense> = emptyList(),
)
